src/metabase/api/cloud_migration.clj

```diff
@@ -0,0 +1,204 @@
+(ns metabase.api.cloud-migration
+  "/api/cloud-migration endpoints.
+  Only one migration should be happening at any given time.
+  But if something weird happens with concurrency, /cancel will
+  cancel all of them. "
+  (:require
+   [cheshire.core :as json]
+   [clj-http.client :as http]
+   [clojure.java.io :as io]
+   [clojure.set :as set]
+   [compojure.core :refer [GET POST PUT]]
+   [metabase.api.common :as api]
+   [metabase.cmd.dump-to-h2 :as dump-to-h2]
+   [metabase.config :as config]
+   [metabase.models.cloud-migration :refer [CloudMigration]]
+   [metabase.models.setting :refer [defsetting]]
+   [metabase.models.setting.cache :as setting.cache]
+   [metabase.util.i18n :refer [deferred-tru]]
+   [metabase.util.log :as log]
+   [methodical.core :as methodical]
+   [toucan2.core :as t2]
+   [toucan2.pipeline :as t2.pipeline])
+  (:import
+   [java.io File FileInputStream InputStream]))
+
+(set! *warn-on-reflection* true)
+
+(def ^:private ^String metabase-store-migration-url
+  "https://store-api.metabase.com/api/v2/migration")
+
+(def ^:private ^java.io.File dump-file
+  (io/file "cloud_migration_dump.db"))
+
+;; Read-Only mode
+
+(defsetting read-only-mode
+  (deferred-tru
+    (str "Boolean indicating whether a Metabase's is in read-only mode with regards to its app db. "
+         "Will take up to 1m to propagate to other Metabase instances in a cluster."
+         "Audit tables are excluded from read-only-mode mode."))
+  :type       :boolean
+  :visibility :admin
+  :default    false
+  :doc        false
+  :export?    false)
+
+(def ^:private read-only-mode-exceptions
+  #{;; Migrations need to update their own state
+    :model/CloudMigration
+    :model/Setting
+
+    ;; Users need to login, make queries, and we need need to audit them.
+    :model/Session
+    :model/User
+    :model/LoginHistory
+    :model/AuditLog
+    :model/QueryExecution
+    :model/ViewLog
+
+    ;; TODO: Came up in tests, check with core folks
+    :setting
+    :permissions_group_membership})
+
+;; Block write calls to most tables in read-only mode.
+(methodical/defmethod t2.pipeline/build :before [#_query-type     :toucan.statement-type/DML
```

Comment:
@camsaul PTAL if this is what you had in mind. I used a set instead of a `derive` like you suggested, but maybe I missed why that was important. I'd also appreciate your input on the `read-only-mode-exceptions`.